pkg/download/bucket/bucket.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/idutils"
+	jsonutils "github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/json"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/internal/log/field"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/coordinate"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/parameter"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/parameter/value"
+	"github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/config/template"
+	v2 "github.com/dynatrace/dynatrace-configuration-as-code/v2/pkg/project/v2"
+	"strings"
+)
+
+type BucketClient interface {
+	List(ctx context.Context) (buckets.ListResponse, error)
+}
+
+type Downloader struct {
+	client BucketClient
+}
+
+func NewDownloader(client BucketClient) *Downloader {
+	return &Downloader{
+		client: client,
+	}
+}
+func (d *Downloader) Download(projectName string, _ ...config.BucketType) (v2.ConfigsPerType, error) {
+	result := make(v2.ConfigsPerType)
+	response, err := d.client.List(context.TODO())
+	if err != nil {
+		log.WithFields(field.Type("bucket"), field.Error(err)).Error("Failed to fetch all bucket definitions: %v", err)
+		return result, nil
+	}
+
+	if !response.IsSuccess() {
+		log.WithFields(field.Type("bucket"), field.Error(err)).Error("Failed to fetch all bucket definitions: %s", string(response.Data))
+		return result, nil
+	}
+
+	configs := d.convertAllObjects(projectName, response.Objects)
+	result["bucket"] = configs
+	return result, nil
+}
+
+func (d *Downloader) convertAllObjects(projectName string, objects [][]byte) []config.Config {
+	result := make([]config.Config, 0, len(objects))
+	for _, o := range objects {
+		var bucketName struct {
+			BucketName  string `json:"bucketName"`
+			DisplayName string `json:"displayName"`
+		}
+		if err := json.Unmarshal(o, &bucketName); err != nil {
+			return result
+		}
+
+		// exclude builtin bucket names
+		if strings.HasPrefix(bucketName.BucketName, "dt_") || strings.HasPrefix(bucketName.BucketName, "default_") {
+			continue
+		}
+
+		// construct config object with generated config ID
+		configID := idutils.GenerateUUIDFromString(bucketName.BucketName)
+		configCoord := coordinate.Coordinate{
+			Project:  projectName,
+			Type:     "bucket",
+			ConfigId: configID,
+		}
+
+		params := map[string]parameter.Parameter{}
+		{
+			var s string
+			var err error
+
+			s, err = getValueForAttribute(o, "bucketName")
+			if err != nil {
+				log.WithFields(field.Coordinate(coordinate.Coordinate{Project: projectName, Type: "bucket", ConfigId: configID}), field.Error(err)).Warn("Failed to get configuration for %v (%s): %v", configID, "bucket", err)
+				continue
+			}
+			params[config.NameParameter] = &value.ValueParameter{Value: s}
```

Comment:
Up for discussion, but I would **not** set the `bucketName` as the `name` parameter!
I would still extract it but into a `bucketName` parameter, rather than the reserved `NameParameter`.

We don't use the `name` to identify buckets, which already caused some confusion for internal preview users. 

As a user, you have a config `ID`, a config `name`, and the `originObjectId` for an API that uses `bucketName` as an identifier.
And you probably know that at least config API types are referenced by name. So seeing a bucket config with a `name`, very reasonable leads to the assumption that this will be the (bucket)Name.

Reply:
Instead of the config's name, should we start using a custom parameter, e.g. "displayName"? I think it's more natural to use these than the magic "name" field. The name field was used more for the config's name that we use to look up ids, here it's a simple displayname.

Reply:
+1 and we've already **exactly** agreed on doing it that way :D  (skip name, introduce displayName)

Reply:
Slight correction to what both of you said @Laubi @warber: we need to have two parameters, `bucketName` and `displayName` (we already do that line 115) are separate things.  
 So this one should just be written to a bucketName param, not the reserved name

Reply:
The `bucketName` should be written into the `originObjectId`, right? As for reupload on the same env, it should update the existing objects

Reply:
Hmm.. fair point. And we do write it into the JSON payload on upload as well right?
In that case we should be removing it from the JSON on download, not extracting it into a templating parameter, that's right!